Use a byte for OperationType

Geod24 opened the pull request:
```
Instead of storing a string, which takes up quite some space,
we store a single byte and provide a mean to convert to/from string.
```

kfangw commented:
[transaction_extract_test.txt](https://github.com/bosnet/sebak/files/2620177/transaction_extract_test.txt)
Maybe this is helpful for extracting the signing value.
@Geod24

Charleslee522 commented on lib/node/runner/api/operation.go:
```diff
@@ -24,22 +24,24 @@ func (api NetworkHandlerAPI) GetOperationsByAccountHandler(w http.ResponseWriter
-		httputils.WriteJSONError(w, errors.InvalidQueryString)
-		return
+	if len(oTypeStr) > 0 {
+		if err = oType.UnmarshalText([]byte(oTypeStr)); err != nil {
```
way better 👍

Charleslee522 commented:
The reason of test failure in `integration_tests_node` and `long_term` is because of network disconnect. I'll check it more.
``` integration_tests_node
t=2018-11-27T11:34:47+0000 lvl=dbug msg="validator is disconnected" module=network node=GDIR.M6CC validator=GCDCXYUTLFOZSRQ4K6DTZ3R7ZJMD6CHVL3ZM7KGOG52NOTNHWBKYPCIO error="Post https://127.0.0.1:2824/node/connect: dial tcp 127.0.0.1:2824: connect: connection refused" caller=validator_connection_manager.go:160
```
``` long_term
t=2018-11-27T11:34:50+0000 lvl=dbug msg="validator is disconnected" module=network node=GDIR.M6CC validator=GCDCXYUTLFOZSRQ4K6DTZ3R7ZJMD6CHVL3ZM7KGOG52NOTNHWBKYPCIO error="Post https://127.0.0.1:2824/node/connect: dial tcp 127.0.0.1:2824: connect: connection refused" caller=validator_connection_manager.go:160
```

kfangw commented:
@Geod24 Changes looks good.
If the CI pass, I will review again.

Geod24 commented:
I would prefer #826 to be merged first. I tried to get this in before MainNet, as it can potentially break the database, but couldn't. So I will have to look into #834 / #842 before we merge this.

Geod24 commented:
I rebased on master, and reworked my approach.
In order for us not to break MainNet compatibility, we cannot have the hash change, since that's what the signature is based on. We don't want valid transaction to start being rejected.

In order to do so, I had to tweak the way the hashing is done, or more precisely, the encoding.
The hashing is done on the encoded RLP representation.
Something like:

```go
type Header struct {
  Type string
}
```

Is encoded as a list containing a single string element. Any aggregate (`struct`) adds a layer of list.
There are some special rules for list length which are explained [here](https://github.com/ethereum/wiki/wiki/RLP). In order to keep the hash the same, I had to keep the same encoding, even if it is wasteful.

So, this patches now only helps with memory. Instead of storing a string for each operation, only a `byte` will be stored. Once this gets approved, I will submit a patch for an upgrade path for the DB.

@spikeekips @Charleslee522 @kfangw please re-review as a lot of code got added.
If it's too big, I can split the PR in two (as the first few commits are just refactoring).

kfangw commented:
Except for the `hash.go`, everything looks good.

Geod24 commented on lib/common/hash.go:
```diff
@@ -1,18 +1,140 @@
+}
+
+// Encodes a list to RLP, including its size
+func EncodeList(items ...interface{}) ([]byte, error) {
```
The function you linked is private, so I cannot use it.
And it does not provide the same kind of functionality.

As I said in [this message](https://github.com/bosnet/sebak/pull/836#issuecomment-444562601), the header gets encoded as a list with a single element. This function allows to encode list, and no public function in the go-ethereum code lets me do that.

I am considering making a PR to go-ethereum to make some stuff available and remove a bunch of code from here, but I did not want to block this PR on another project.

kfangw replied:
https://github.com/ethereum/go-ethereum/blob/3ac633ba8423851a85cb909a1f1f0f3c0d2c5351/rlp/encode.go#L80-L93
By using the function `func Encode(w io.Writer, val interface{}) error`,
We can encode an object according to its type.

But I understand your intention.
If there is a test suite that ensures the functions work well, it will be good.
How do you think about this?

Geod24 replied:
Indeed, but I want to encode the object without having to make an extra type.
I'll add tests for it.

anarcher replied:
Why does `EncodeList` need?  As told @kfangw ,  By using `Encode(w,v) e` , we can encode any object even string or int. I am not sure but `EncodeList` has weak compatibility with outside sebak/go

Geod24 replied:
Calling `Encode` will only encode one object, while `EncodeList` encodes multiple object into a list.
So in pseudo-code:
```
var a, b, c
EncodeList(a, b, c) => [a, b, c]
Encode(a); Encode(b); Encode(c) => [a] [b] [c]
```

That's why I needed the extra function

anarcher replied:
type xs []string
...
xs.Encode(...) 

It doesn’t works?

Geod24 replied:
```go
func TestAnarcher (t *testing.T) {
    if r1, e1 := EncodeList(uint(15), uint(9), uint(2)); e1 != nil {
        panic(e1)
    } else {
        fmt.Println(r1)
    }
    var buff bytes.Buffer
    w := bufio.NewWriter(&buff)
    Encode(w, uint(15))
    Encode(w, uint(9))
    Encode(w, uint(2))
    w.Flush()
    fmt.Println(buff.Bytes())
    t.FailNow()
}
```

```
[195 15 9 2]
[15 9 2]
```

The first one is a list, the second one just consecutive items.

Geod24 replied:
Actually there is a much simpler way to do this: https://github.com/bosnet/sebak/pull/903
